handlers/handbook.go

```diff
@@ -0,0 +1,159 @@
+	"regexp"
+	"net/http"
+	"html"
+	"io/ioutil"
+	"github.com/bwmarrin/discordgo"
+	"github.com/unswpcsoc/PCSocBot/commands"
+)
+
+var (
+
+	ErrNoInput = errors.New("No Course code entered")
+	ErrInvalidFormat = errors.New("Invalid Course Code Format")
+	ErrNotFound = errors.New("Course Not Found")
+)
+
+type handbook struct {
+	nilCommand
+	Code string `arg:"code"`
+}
+
+func newHandbook() *handbook { return &handbook{} }
+
+func (h *handbook) Aliases() []string { return []string{"handbook"} }
+
+func (h *handbook) Desc() string { return "Searches handbook.unsw for course" }
+
+func (h *handbook) MsgHandle(ses *discordgo.Session, msg *discordgo.Message) (*commands.CommandSend, error) {
+
+	// Special case for DELL1234
+	if strings.ToUpper(h.Code) == "DELL1234" {
+		message := commands.NewSend(msg.ChannelID)
+		embed := makeMessage("https://webapps.cse.unsw.edu.au/webcms2/course/index.php?cid=1137",
+				"How to blow up my computer?",
+				"No Course Outline (yet)",
+				"None",
+				"None")
+		return message.Embed(embed), nil
+	}
+
+	// Check course code
+	match, err := regexp.MatchString(`^[A-Za-z]{4}[0-9]{4}$`, h.Code)
+	if err != nil {
+		return nil, err
+	}
+	if !match {
+		return nil, ErrInvalidFormat
+	}
+
+	//assume undergraduate
+	url, htmltext, status, err := getResponse(h.Code, "undergraduate")
+	if err != nil {
+		return nil, err
+	}
+
+	//if not undergraduate then must be postgraduate
+	if status != 200 {
+		url, htmltext, status, err = getResponse(h.Code, "postgraduate")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if status != 200 {
+		// probably a 404 error but whatever it is, the page could not be accessed
+		return nil, ErrNotFound
+	}
+
+	// initialise info
+	title, desc, term, cond := "", "", "", ""
+
+	// extract info via shifty means
+	lines := bytes.Split(htmltext, []byte("\n"))
+	for i, line := range lines {
+		line := string(line[:])
+		if title == "" {
+			if strings.Contains(line, "module-title") {
+				title = strings.Split(strings.Split(line, ">")[2], "<")[0]
+			}
+
+		}
+		if desc == "" {
+			if strings.Contains(line, "readmore__wrapper"){
+				desc = strings.Split(strings.Split(string(lines[i+2][:]), ">")[1], "<")[0]
+				desc = html.UnescapeString(desc)
+			}
+		}
+		if term == "" {
+			if strings.Contains(line, "<p tabindex=\"0\" class=\"\">"){
+				term = strings.Split(strings.Split(string(line), ">")[1], "<")[0]
+			}
+		}
+		if cond == "" {
+			if strings.Contains(line, "Prerequisite") {
+				cond = strings.Split(strings.Split(string(line), ">")[1], "<")[0]
+			}
+		}
+		if title != "" && desc != "" && term != "" && cond != "" {
+			// stop early if all fields are filled
+			break
+		}
+	}
+	// If no prerequisites found
+	if cond == "" {
+		cond = "None"
+	}
+
+	// create and send message
+	message := commands.NewSend(msg.ChannelID)
+	embed := makeMessage(url, title, desc, term, cond)
+	return message.Embed(embed), nil
+}
+
+
+func getResponse(Code string, Graduate string) (string, []byte, int, error) {
```

Comment:
In this case I would either write a function annotation or use named return values to make the function's... functionality more clear.

Reply:
Done